File: src/test/java/org/frc1778/freezylib/logging/FreezyLogTest.java
package org.frc1778.freezylib.logging;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.assertThatCode;
import static org.assertj.core.api.Assertions.assertThatExceptionOfType;
import static org.junit.jupiter.api.Assertions.fail;

import com.google.common.io.Files;
import edu.wpi.first.wpilibj.DriverStation.MatchType;
import java.io.File;
import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.nio.file.Path;
import org.frc1778.freezylib.util.Measurement;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.io.TempDir;

/**
 * Tests the FreezyLog class and main methods.
 *
 * @author FRC 1778 Chill Out
 */
public class FreezyLogTest {

  @BeforeEach
  public void dumpFreezyLog() {
    FreezyLog.dump();
  }

  @Test
  public void providingAnInvalidFileNameShouldThrowException() {
    assertThatExceptionOfType(Exception.class)
        .isThrownBy(() -> FreezyLog.setMetaFileName("/meta.notjson"));
    assertThatCode(() -> FreezyLog.setMetaFileName("/meta.json")).doesNotThrowAnyException();

    assertThatExceptionOfType(Exception.class)
        .isThrownBy(() -> FreezyLog.setLogFileName("/log.notcsv"));
    assertThatCode(() -> FreezyLog.setLogFileName("/log.csv")).doesNotThrowAnyException();
  }

  @Test
  public void freezyLogShouldFormatCsvAndJsonFilesToMatchExamples(@TempDir Path folder) {
    FreezyLog.setPath(folder.toString());
    FreezyLog.populateMatchStructure(MatchType.Qualification, 0);
    FreezyLog.addField(new MetaField("MetaField", Measurement.Unitless.UNITLESS, "meta"));
    FreezyLog.addField(new PolledField("PolledField", Measurement.Length.INCHES, () -> "polled"));
    var subscribedField =
        new SubscribedField("SubscribedField", Measurement.Angle.DEGREES, String.class);
    FreezyLog.addField(subscribedField);
    subscribedField.pushValue("subscribed");

    FreezyLog.log();

    try {
      assertThat(readFileAsCharSource(FreezyLog.getLogFile()))
          .isEqualTo(
              readFileAsCharSource(
                  new File(
                      Thread.currentThread()
                          .getContextClassLoader()
                          .getResource("log.csv")
                          .getFile())));
      assertThat(readFileAsCharSource(FreezyLog.getMetaFile()))
          .isEqualTo(
              readFileAsCharSource(
                  new File(
                      Thread.currentThread()
                          .getContextClassLoader()
                          .getResource("meta.json")
                          .getFile())));
    } catch (IOException e) {
      fail(e.getMessage());
    }
  }

  private String readFileAsCharSource(File file) throws IOException {
    return Files.asCharSource(file, StandardCharsets.UTF_8)
        .read()
        .replaceAll("\\r\\n", "\n")
        .replaceAll("\\r", "\n");
  }
}

File: src/test/kotlin/org/frc1778/freezylib/logging/FreezyLogTest.kt
package org.frc1778.freezylib.logging

import edu.wpi.first.wpilibj.DriverStation.MatchType
import io.kotlintest.shouldBe
import io.kotlintest.shouldNotThrowAny
import io.kotlintest.shouldThrow
import io.kotlintest.specs.StringSpec
import java.io.IOException
import org.frc1778.freezylib.util.Measurement

/**
 * Tests the FreezyLog class and main methods.
 *
 * @author FRC 1778 Chill Out
 */
class FreezyLogTest : StringSpec() {

    init {
        FreezyLog.dump()
        FreezyLog.setPath(createTempDir().toString())

        "An invalid file name should throw an exception" {
            shouldThrow<Exception> { FreezyLog.setMetaFileName("/meta.notjson") }
            shouldThrow<Exception> { FreezyLog.setLogFileName("/meta.notcsv") }
        }

        "A valid file name should not throw an exception" {
            shouldNotThrowAny { FreezyLog.setMetaFileName("/meta.json") }
            shouldNotThrowAny { FreezyLog.setLogFileName("/meta.csv") }
        }

        FreezyLog.populateMatchStructure(MatchType.Qualification, 0)
        FreezyLog.addField(MetaField("MetaField", Measurement.Unitless.UNITLESS, "meta"))
        FreezyLog.addField(PolledField("PolledField", Measurement.Length.INCHES) { "polled" })

        val taggedField = TaggedField("TaggedField", Measurement.Acceleration.METERS_PER_SECOND_SQUARED, Int::class)
        taggedField.updateTag(0)
        FreezyLog.addField(taggedField)

        val subscribedField = SubscribedField("SubscribedField", Measurement.Angle.DEGREES, Int::class)
        subscribedField.pushValue(0)
        FreezyLog.addField(subscribedField)

        try {
            throw IOException("Forcefully thrown exception")
        } catch (e: Exception) {
            FreezyLog.logException(e)
        }

        for (i in 1 until 4) {
            FreezyLog.log()
            subscribedField.pushValue(i)
            taggedField.updateTag(-i)
        }

        "Log file should match with example CSV" {
            FreezyLog.logFile.readText() shouldBe Thread.currentThread().contextClassLoader.getResource("log.csv")!!.readText()
        }
        "Meta file should match with example JSON" {
            FreezyLog.metaFile.readText() shouldBe Thread.currentThread().contextClassLoader.getResource("meta.json")!!.readText()
        }
    }
}
